File: balanced_binary_tree.py

class Node(object):
    def __init__(self,val):
        self.val = val
        self.left = None
        self.right = None

class Solution:
    def balanced_helper(self,root):
        if root is None:
            return 0
        l = self.balanced_helper(root.left)
        r = self.balanced_helper(root.right)
        if l<0 or r<0 or abs(l-r)>1:
            return -1
        return max(l,r)+1
    # @param root, a tree node
    # @return a boolean
    def isBalanced(self, root):
        return True if self.balanced_helper(root)>=0 else False


if __name__ == "__main__":
    n1 = Node(1)
    n2 = Node(2)
    n3 = Node(3)
    n4 = Node(4)
    n5 = Node(5)
    n6 = Node(6)
    n1.left = n2
    n1.right = n5
    n2.left = n3
    n2.right = n4
    n5.right = n6
    print Solution().isBalanced(n1)

File: balanced_binary_tree.cpp
/*
 *
Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1. 

*/

#include  "header.h"

bool isBalanced(TreeNode* root, int& height){
    if(!root){
        height = 0;
        return true;
    }
    int l_h = 0, r_h=0;
    bool res = isBalanced(root->left, l_h) && isBalanced(root->right, r_h);
    height = max(l_h, r_h)+1;
    return res && abs(l_h-r_h)<=1;
}

bool isBalanced(TreeNode* root){
    if(!root) return true;
    int height = 0;
    return isBalanced(root, height);
}

int main(){
    TreeNode* root = create_tree("12534#1");
    printf("%d\n", isBalanced(root));
    return 0;
}
